packages/token/src/common/erc2981/erc2981.cairo

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.15.1 (token/common/erc2981/erc2981.cairo)
+
+/// # ERC2981 Component
+///
+/// Implementation of the NFT Royalty Standard, a standardized way to retrieve royalty payment
+/// information.
+///
+/// Royalty information can be specified globally for all token ids via `set_default_royalty`,
+/// and/or individually for specific token ids via `set_token_royalty`. The latter takes precedence
+/// over the first.
+///
+/// Royalty is specified as a fraction of sale price. The denominator is set by the contract by
+/// using the Immutable Component Config pattern.
+/// See https://community.starknet.io/t/immutable-component-config/114434.
+///
+/// IMPORTANT: ERC-2981 only specifies a way to signal royalty information and does not enforce its
+/// payment. See https://eips.ethereum.org/EIPS/eip-2981#optional-royalty-payments[Rationale] in the
+/// ERC. Marketplaces are expected to voluntarily pay royalties together with sales.
+#[starknet::component]
+pub mod ERC2981Component {
+    use core::num::traits::Zero;
+    use openzeppelin_introspection::src5::SRC5Component::InternalTrait as SRC5InternalTrait;
+    use openzeppelin_introspection::src5::SRC5Component::SRC5Impl;
+    use openzeppelin_introspection::src5::SRC5Component;
+    use openzeppelin_token::common::erc2981::interface::{IERC2981, IERC2981_ID};
+    use starknet::ContractAddress;
+    use starknet::storage::Map;
+
+    // This default denominator is only used when the DefaultConfig
+    // is in scope in the implementing contract.
+    pub const DEFAULT_FEE_DENOMINATOR: u128 = 10_000;
+
+    #[derive(Serde, Drop, starknet::Store)]
+    struct RoyaltyInfo {
+        pub receiver: ContractAddress,
+        pub royalty_fraction: u128,
+    }
+
+    #[storage]
+    struct Storage {
+        default_royalty_info: RoyaltyInfo,
+        token_royalty_info: Map<u256, RoyaltyInfo>,
+    }
+
+    mod Errors {
+        pub const INVALID_ROYALTY: felt252 = 'ERC2981: invalid royalty';
+        pub const INVALID_ROYALTY_RECEIVER: felt252 = 'ERC2981: invalid receiver';
+        pub const INVALID_FEE_DENOMINATOR: felt252 = 'Invalid fee denominator';
+    }
+
+    /// Constants expected to be defined at the contract level used to configure the component
+    /// behaviour.
+    ///
+    /// - `FEE_DENOMINATOR`: The denominator with which to interpret the fee set in
+    ///   `set_token_royalty` and `set_default_royalty` as a fraction of the sale price.
+    ///
+    /// Requirements:
+    ///
+    /// - FEE_DENOMINATOR must be greater than 0.
+    pub trait ImmutableConfig {
+        const FEE_DENOMINATOR: u128;
+
+        fn validate() {
+            assert(Self::FEE_DENOMINATOR > 0, Errors::INVALID_FEE_DENOMINATOR);
```

Comment:
Mmm I don't love this. I get the idea, but I think it's hard to justify including a testing-only fn in the trait tbh. Since the assertion is included in the default fn, this could get weird if a user decides to define their own `validate` fn for whatever reason when creating their own impl (even if it's documented that users shouldn't do this). Including an assertion in the `initializer` with the error msg IMO is much more straight forward. Do you agree?

Also, I wasn't sure if posting this here or in the shamans proposal would be better. I can move this comment over there to further discuss if you'd like

Reply:
> Agree the testing only function is not the best approach, but that is not the intention of validate anymore

Ahh I see. I read the shamans post while going over this and thought that was still the intention

> first one, the initializer grows big if we have a lot of config assertions

Yeah, this is true. I'm wondering if it's more or less readable and explicit when the assertions are abstracted away in the trait vs the `initializer`

> second one is that the component may not have an initializer

This is a good point

----

My main concern with this idea is that it's really easy to create the immutable impl and improperly redefine the `validate` fn. Do you see any way of mitigating this (aside from being super loud in the docs)?

Reply:
I don't see a way around it. I think that should be part of the specification (SNIP).

Reply:
[SRC-107](https://github.com/starknet-io/SNIPs/blob/963848f0752bde75c7087c2446d83b7da8118b25/SNIPS/snip-107.md)

Reply:
Maybe we can separate the immutable constants and `validate` with supertraits...we've been talking about supertraits for forever now :( it might be worth thinking about if/when they're are a thing

Also, nice work on the src!